File: backend-rust/tests/support/tenant_fixture.rs
use std::collections::BTreeMap;

use serde_json::{json, Value};

pub const TENANT_A_ID: &str = "20001";
pub const TENANT_B_ID: &str = "20002";
pub const ACTOR_USER_ID: &str = "10001";
pub const REQUEST_ID: &str = "8b9c0e7f-9d8b-4f86-b77d-9a3d4c2f5001";
pub const OCCURRED_AT: &str = "2026-08-06T13:01:00.123Z";

pub const REQUIRED_INTERNAL_HEADERS: [&str; 5] = [
    "X-Tenant-Id",
    "X-User-Id",
    "X-Request-Id",
    "X-Internal-Timestamp",
    "X-Internal-Signature",
];

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Tenant {
    pub id: String,
    pub code: String,
    pub role: String,
    pub permissions: Vec<String>,
}

impl Tenant {
    pub fn new(id: &str, code: &str, role: &str, permissions: &[&str]) -> Self {
        Self {
            id: id.to_string(),
            code: code.to_string(),
            role: role.to_string(),
            permissions: permissions.iter().map(|value| (*value).to_string()).collect(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Resource {
    pub id: String,
    pub kind: String,
    pub tenant_id: String,
    pub parent_id: Option<String>,
    pub storage_path: String,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Page {
    pub tenant_id: String,
    pub query: String,
    pub page: u32,
    pub size: u32,
    pub total: u64,
    pub items: Vec<Resource>,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct DownloadPreview {
    pub resource_id: String,
    pub tenant_id: String,
    pub download_path: String,
    pub preview_path: String,
}

#[derive(Clone, Debug, PartialEq)]
pub struct SseEvent {
    pub schema_version: u8,
    pub event_id: String,
    pub event: String,
    pub tenant_id: String,
    pub task_id: String,
    pub occurred_at: String,
    pub data: Value,
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct QueueMessage {
    pub schema_version: u8,
    pub tenant_id: String,
    pub task_id: String,
    pub actor_user_id: String,
    pub request_id: String,
}

impl QueueMessage {
    pub fn as_json(&self) -> Value {
        json!({
            "schema_version": self.schema_version,
            "tenant_id": self.tenant_id,
            "task_id": self.task_id,
            "actor_user_id": self.actor_user_id,
            "request_id": self.request_id,
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct InternalRequest {
    pub method: String,
    pub path_with_query: String,
    pub body_sha256: String,
    pub timestamp: String,
    pub tenant_id: String,
    pub user_id: String,
    pub request_id: String,
    pub signature: String,
    pub replay_key: String,
    pub canonical_request: String,
    pub headers: BTreeMap<String, String>,
}

#[derive(Clone, Debug)]
pub struct TenantFixture {
    pub actor_user_id: String,
    pub tenant_a: Tenant,
    pub tenant_b: Tenant,
    pub resources: Vec<Resource>,
    pub pages: Vec<Page>,
    pub download_previews: Vec<DownloadPreview>,
    pub events: Vec<SseEvent>,
    pub queue_messages: Vec<QueueMessage>,
    pub internal_requests: Vec<InternalRequest>,
}

impl TenantFixture {
    pub fn builder() -> TenantFixtureBuilder {
        TenantFixtureBuilder::default()
    }

    pub fn resource_for(&self, tenant_id: &str, resource_id: &str) -> &Resource {
        self.resources
            .iter()
            .find(|resource| resource.tenant_id == tenant_id && resource.id == resource_id)
            .expect("fixture resource must exist")
    }

    pub fn page_for(&self, tenant_id: &str) -> &Page {
        self.pages
            .iter()
            .find(|page| page.tenant_id == tenant_id)
            .expect("fixture page must exist")
    }

    pub fn download_preview_for(&self, tenant_id: &str) -> &DownloadPreview {
        self.download_previews
            .iter()
            .find(|case| case.tenant_id == tenant_id)
            .expect("fixture download/preview case must exist")
    }

    pub fn event_for(&self, tenant_id: &str) -> &SseEvent {
        self.events
            .iter()
            .find(|event| event.tenant_id == tenant_id)
            .expect("fixture SSE event must exist")
    }

    pub fn queue_message_for(&self, tenant_id: &str) -> &QueueMessage {
        self.queue_messages
            .iter()
            .find(|message| message.tenant_id == tenant_id)
            .expect("fixture queue message must exist")
    }

    pub fn internal_request_for(&self, tenant_id: &str) -> &InternalRequest {
        self.internal_requests
            .iter()
            .find(|request| request.tenant_id == tenant_id)
            .expect("fixture internal request must exist")
    }
}

#[derive(Clone, Debug)]
pub struct TenantFixtureBuilder {
    actor_user_id: String,
    tenant_a: Tenant,
    tenant_b: Tenant,
}

impl Default for TenantFixtureBuilder {
    fn default() -> Self {
        Self {
            actor_user_id: ACTOR_USER_ID.to_string(),
            tenant_a: Tenant::new(
                TENANT_A_ID,
                "acme-bid",
                "ADMIN",
                &["tenant.read", "tender.write", "audit.start", "audit.report.read"],
            ),
            tenant_b: Tenant::new(
                TENANT_B_ID,
                "solo-10001",
                "OWNER",
                &["tenant.read", "tenant.delete", "audit.report.read"],
            ),
        }
    }
}

impl TenantFixtureBuilder {
    pub fn actor_user_id(mut self, actor_user_id: &str) -> Self {
        self.actor_user_id = actor_user_id.to_string();
        self
    }

    pub fn tenant_a(mut self, tenant: Tenant) -> Self {
        self.tenant_a = tenant;
        self
    }

    pub fn tenant_b(mut self, tenant: Tenant) -> Self {
        self.tenant_b = tenant;
        self
    }

    pub fn build(self) -> TenantFixture {
        let project_a = resource(&self.tenant_a, "project-01", "project", None);
        let document_a = resource(
            &self.tenant_a,
            "document-01",
            "document",
            Some(project_a.id.clone()),
        );
        let project_b = resource(&self.tenant_b, "project-01", "project", None);
        let document_b = resource(
            &self.tenant_b,
            "document-01",
            "document",
            Some(project_b.id.clone()),
        );

        let pages = vec![
            Page {
                tenant_id: self.tenant_a.id.clone(),
                query: "smart campus".to_string(),
                page: 1,
                size: 20,
                total: 1,
                items: vec![document_a.clone()],
            },
            Page {
                tenant_id: self.tenant_b.id.clone(),
                query: "smart campus".to_string(),
                page: 1,
                size: 20,
                total: 1,
                items: vec![document_b.clone()],
            },
        ];
        let download_previews = vec![
            download_preview(&document_a),
            download_preview(&document_b),
        ];
        let events = vec![
            event(&self.tenant_a, "task-01", "1002"),
            event(&self.tenant_b, "task-01", "1002"),
        ];
        let queue_messages = vec![
            queue_message(&self.tenant_a, &self.actor_user_id, "task-01"),
            queue_message(&self.tenant_b, &self.actor_user_id, "task-01"),
        ];
        let internal_requests = vec![
            internal_request(&self.tenant_a, &self.actor_user_id),
            internal_request(&self.tenant_b, &self.actor_user_id),
        ];

        TenantFixture {
            actor_user_id: self.actor_user_id,
            tenant_a: self.tenant_a,
            tenant_b: self.tenant_b,
            resources: vec![project_a, document_a, project_b, document_b],
            pages,
            download_previews,
            events,
            queue_messages,
            internal_requests,
        }
    }
}

fn resource(tenant: &Tenant, id: &str, kind: &str, parent_id: Option<String>) -> Resource {
    Resource {
        id: id.to_string(),
        kind: kind.to_string(),
        tenant_id: tenant.id.clone(),
        parent_id,
        storage_path: format!(
            "storage/tenants/{}/tenders/2026-08-06/{}.pdf",
            tenant.id, id
        ),
    }
}

fn download_preview(resource: &Resource) -> DownloadPreview {
    DownloadPreview {
        resource_id: resource.id.clone(),
        tenant_id: resource.tenant_id.clone(),
        download_path: format!("/api/resources/{}/download", resource.id),
        preview_path: format!("/api/resources/{}/preview", resource.id),
    }
}

fn event(tenant: &Tenant, task_id: &str, event_id: &str) -> SseEvent {
    SseEvent {
        schema_version: 1,
        event_id: event_id.to_string(),
        event: "progress".to_string(),
        tenant_id: tenant.id.clone(),
        task_id: task_id.to_string(),
        occurred_at: OCCURRED_AT.to_string(),
        data: json!({
            "stage": "legal",
            "percent": 42,
            "message": "reviewing",
        }),
    }
}

fn queue_message(tenant: &Tenant, actor_user_id: &str, task_id: &str) -> QueueMessage {
    QueueMessage {
        schema_version: 1,
        tenant_id: tenant.id.clone(),
        task_id: task_id.to_string(),
        actor_user_id: actor_user_id.to_string(),
        request_id: REQUEST_ID.to_string(),
    }
}

fn internal_request(tenant: &Tenant, actor_user_id: &str) -> InternalRequest {
    let method = "POST".to_string();
    let path_with_query = "/api/v1/review/document-01/stream?mode=full".to_string();
    let timestamp = "1786019460".to_string();
    let body_sha256 = "a".repeat(64);
    let request_id = REQUEST_ID.to_string();
    let signature = format!("v1={}", "b".repeat(64));
    let canonical_request = [
        "v1",
        method.as_str(),
        path_with_query.as_str(),
        timestamp.as_str(),
        tenant.id.as_str(),
        actor_user_id,
        request_id.as_str(),
        body_sha256.as_str(),
    ]
    .join("\n");
    let mut headers = BTreeMap::new();
    headers.insert("X-Tenant-Id".to_string(), tenant.id.clone());
    headers.insert("X-User-Id".to_string(), actor_user_id.to_string());
    headers.insert("X-Request-Id".to_string(), request_id.clone());
    headers.insert("X-Internal-Timestamp".to_string(), timestamp.clone());
    headers.insert("X-Internal-Signature".to_string(), signature.clone());

    InternalRequest {
        method,
        path_with_query,
        body_sha256,
        timestamp,
        tenant_id: tenant.id.clone(),
        user_id: actor_user_id.to_string(),
        request_id: request_id.clone(),
        signature,
        replay_key: format!("{}:{}", tenant.id, request_id),
        canonical_request,
        headers,
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct NegativeCase {
    pub name: &'static str,
    pub surface: &'static str,
    pub activation: &'static str,
    pub expected_error_code: &'static str,
}

pub const NEGATIVE_CASES: &[NegativeCase] = &[
    NegativeCase {
        name: "cross_tenant_id_lookup_is_not_visible",
        surface: "resource lookup by tenantId and resourceId",
        activation: "T3",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
    NegativeCase {
        name: "cross_tenant_page_and_search_do_not_mix_items",
        surface: "pagination and search",
        activation: "T3",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
    NegativeCase {
        name: "cross_tenant_parent_child_lookup_is_not_visible",
        surface: "parent-child resource traversal",
        activation: "T3",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
    NegativeCase {
        name: "cross_tenant_download_is_not_visible",
        surface: "download",
        activation: "T4",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
    NegativeCase {
        name: "cross_tenant_preview_is_not_visible",
        surface: "preview and normalized storage path",
        activation: "T4",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
    NegativeCase {
        name: "cross_tenant_sse_replay_is_not_visible",
        surface: "SSE subscription and Last-Event-ID replay",
        activation: "T5",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
    NegativeCase {
        name: "queue_message_without_matching_tenant_context_is_rejected",
        surface: "queue and outbox envelope",
        activation: "T6",
        expected_error_code: "TENANT_CONTEXT_INVALID",
    },
    NegativeCase {
        name: "rust_header_tenant_mismatch_is_rejected",
        surface: "Java-to-Rust internal request",
        activation: "T7",
        expected_error_code: "INTERNAL_TENANT_MISMATCH",
    },
    NegativeCase {
        name: "replayed_request_id_is_scoped_by_tenant",
        surface: "Rust replay cache",
        activation: "T7",
        expected_error_code: "INTERNAL_REQUEST_REPLAYED",
    },
    NegativeCase {
        name: "all_surfaces_keep_the_same_negative_contract",
        surface: "cross-surface regression sweep",
        activation: "T8",
        expected_error_code: "RESOURCE_NOT_FOUND",
    },
];

File: backend-java/src/test/java/com/ithsd/smart_tender/tenant/fixture/TenantFixture.java
package com.ithsd.smart_tender.tenant.fixture;

import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

/**
 * Reusable two-tenant fixture for contract and isolation tests.
 *
 * <p>The fixture deliberately models data rather than calling an endpoint. That
 * keeps the T0 tests runnable before the T1-T8 implementation slices land.</p>
 */
public final class TenantFixture {

    public static final String TENANT_A_ID = "20001";
    public static final String TENANT_B_ID = "20002";
    public static final String ACTOR_USER_ID = "10001";
    public static final String REQUEST_ID = "8b9c0e7f-9d8b-4f86-b77d-9a3d4c2f5001";
    public static final String OCCURRED_AT = "2026-08-06T13:01:00.123Z";

    public record Tenant(String id, String code, String role, Set<String> permissions) {
        public Tenant {
            Objects.requireNonNull(id);
            Objects.requireNonNull(code);
            Objects.requireNonNull(role);
            permissions = Set.copyOf(permissions);
        }
    }

    public record Resource(
            String id,
            String kind,
            String tenantId,
            String parentId,
            String storagePath
    ) {
    }

    public record Page<T>(
            String tenantId,
            String query,
            int page,
            int size,
            long total,
            List<T> items
    ) {
        public Page {
            items = List.copyOf(items);
        }
    }

    public record DownloadPreview(
            String resourceId,
            String tenantId,
            String downloadPath,
            String previewPath
    ) {
    }

    public record SseEvent(
            int schemaVersion,
            String eventId,
            String event,
            String tenantId,
            String taskId,
            String occurredAt,
            Map<String, Object> data
    ) {
        public SseEvent {
            data = Map.copyOf(data);
        }
    }

    public record QueueMessage(
            int schemaVersion,
            String tenantId,
            String taskId,
            String actorUserId,
            String requestId
    ) {
    }

    public record InternalRequest(
            String method,
            String pathWithQuery,
            String bodySha256,
            String timestamp,
            String tenantId,
            String userId,
            String requestId,
            String signature,
            String replayKey,
            String canonicalRequest,
            Map<String, String> headers
    ) {
        public InternalRequest {
            headers = Map.copyOf(headers);
        }
    }

    private final String actorUserId;
    private final Tenant tenantA;
    private final Tenant tenantB;
    private final Map<String, Resource> resources;
    private final Map<String, Page<Resource>> pages;
    private final Map<String, DownloadPreview> downloadPreviews;
    private final Map<String, SseEvent> events;
    private final Map<String, QueueMessage> queueMessages;
    private final Map<String, InternalRequest> internalRequests;

    private TenantFixture(
            String actorUserId,
            Tenant tenantA,
            Tenant tenantB,
            Map<String, Resource> resources,
            Map<String, Page<Resource>> pages,
            Map<String, DownloadPreview> downloadPreviews,
            Map<String, SseEvent> events,
            Map<String, QueueMessage> queueMessages,
            Map<String, InternalRequest> internalRequests
    ) {
        this.actorUserId = actorUserId;
        this.tenantA = tenantA;
        this.tenantB = tenantB;
        this.resources = Map.copyOf(resources);
        this.pages = Map.copyOf(pages);
        this.downloadPreviews = Map.copyOf(downloadPreviews);
        this.events = Map.copyOf(events);
        this.queueMessages = Map.copyOf(queueMessages);
        this.internalRequests = Map.copyOf(internalRequests);
    }

    public static Builder builder() {
        return new Builder();
    }

    public static TenantFixture defaults() {
        return builder().build();
    }

    public String actorUserId() {
        return actorUserId;
    }

    public Tenant tenantA() {
        return tenantA;
    }

    public Tenant tenantB() {
        return tenantB;
    }

    public Resource resource(Tenant tenant, String resourceId) {
        return resources.get(resourceKey(tenant.id(), resourceId));
    }

    private static String resourceKey(String tenantId, String resourceId) {
        return tenantId + ":" + resourceId;
    }

    public Page<Resource> pageFor(Tenant tenant) {
        return pages.get(tenant.id());
    }

    public DownloadPreview downloadPreviewFor(Tenant tenant) {
        return downloadPreviews.get(tenant.id());
    }

    public SseEvent eventFor(Tenant tenant) {
        return events.get(tenant.id());
    }

    public QueueMessage queueMessageFor(Tenant tenant) {
        return queueMessages.get(tenant.id());
    }

    public InternalRequest internalRequestFor(Tenant tenant) {
        return internalRequests.get(tenant.id());
    }

    public static final class Builder {
        private String actorUserId = ACTOR_USER_ID;
        private Tenant tenantA = new Tenant(
                TENANT_A_ID,
                "acme-bid",
                "ADMIN",
                Set.of("tenant.read", "tender.write", "audit.start", "audit.report.read")
        );
        private Tenant tenantB = new Tenant(
                TENANT_B_ID,
                "solo-10001",
                "OWNER",
                Set.of("tenant.read", "tenant.delete", "audit.report.read")
        );

        public Builder actorUserId(String actorUserId) {
            this.actorUserId = Objects.requireNonNull(actorUserId);
            return this;
        }

        public Builder tenantA(Tenant tenantA) {
            this.tenantA = Objects.requireNonNull(tenantA);
            return this;
        }

        public Builder tenantB(Tenant tenantB) {
            this.tenantB = Objects.requireNonNull(tenantB);
            return this;
        }

        public TenantFixture build() {
            Resource projectA = resource(
                    "project-01",
                    "project",
                    tenantA,
                    null
            );
            Resource documentA = resource(
                    "document-01",
                    "document",
                    tenantA,
                    projectA.id()
            );
            Resource projectB = resource(
                    "project-01",
                    "project",
                    tenantB,
                    null
            );
            Resource documentB = resource(
                    "document-01",
                    "document",
                    tenantB,
                    projectB.id()
            );

            Map<String, Resource> resources = Map.of(
                    resourceKey(tenantA.id(), projectA.id()), projectA,
                    resourceKey(tenantA.id(), documentA.id()), documentA,
                    resourceKey(tenantB.id(), projectB.id()), projectB,
                    resourceKey(tenantB.id(), documentB.id()), documentB
            );
            Map<String, Page<Resource>> pages = Map.of(
                    tenantA.id(), new Page<>(
                            tenantA.id(),
                            "smart campus",
                            1,
                            20,
                            1,
                            List.of(documentA)
                    ),
                    tenantB.id(), new Page<>(
                            tenantB.id(),
                            "smart campus",
                            1,
                            20,
                            1,
                            List.of(documentB)
                    )
            );
            Map<String, DownloadPreview> downloadPreviews = Map.of(
                    tenantA.id(), downloadPreview(documentA),
                    tenantB.id(), downloadPreview(documentB)
            );
            Map<String, SseEvent> events = Map.of(
                    tenantA.id(), event(tenantA, "task-a", "1002"),
                    tenantB.id(), event(tenantB, "task-b", "2002")
            );
            Map<String, QueueMessage> queueMessages = Map.of(
                    tenantA.id(), new QueueMessage(1, tenantA.id(), "task-a", actorUserId, REQUEST_ID),
                    tenantB.id(), new QueueMessage(1, tenantB.id(), "task-b", actorUserId, REQUEST_ID)
            );
            Map<String, InternalRequest> internalRequests = Map.of(
                    tenantA.id(), internalRequest(tenantA),
                    tenantB.id(), internalRequest(tenantB)
            );

            return new TenantFixture(
                    actorUserId,
                    tenantA,
                    tenantB,
                    resources,
                    pages,
                    downloadPreviews,
                    events,
                    queueMessages,
                    internalRequests
            );
        }

        private Resource resource(String id, String kind, Tenant tenant, String parentId) {
            return new Resource(
                    id,
                    kind,
                    tenant.id(),
                    parentId,
                    "storage/tenants/" + tenant.id() + "/tenders/2026-08-06/" + id + ".pdf"
            );
        }

        private DownloadPreview downloadPreview(Resource resource) {
            return new DownloadPreview(
                    resource.id(),
                    resource.tenantId(),
                    "/api/resources/" + resource.id() + "/download",
                    "/api/resources/" + resource.id() + "/preview"
            );
        }

        private SseEvent event(Tenant tenant, String taskId, String eventId) {
            return new SseEvent(
                    1,
                    eventId,
                    "progress",
                    tenant.id(),
                    taskId,
                    OCCURRED_AT,
                    Map.of("stage", "legal", "percent", 42, "message", "reviewing")
            );
        }

        private InternalRequest internalRequest(Tenant tenant) {
            String timestamp = "1786019460";
            String bodySha256 = "a".repeat(64);
            String signature = "v1=" + "b".repeat(64);
            String canonicalRequest = String.join(
                    "\n",
                    "v1",
                    "POST",
                    "/api/v1/review/document-01/stream?mode=full",
                    timestamp,
                    tenant.id(),
                    actorUserId,
                    REQUEST_ID,
                    bodySha256
            );
            Map<String, String> headers = Map.of(
                    "X-Tenant-Id", tenant.id(),
                    "X-User-Id", actorUserId,
                    "X-Request-Id", REQUEST_ID,
                    "X-Internal-Timestamp", timestamp,
                    "X-Internal-Signature", signature
            );
            return new InternalRequest(
                    "POST",
                    "/api/v1/review/document-01/stream?mode=full",
                    bodySha256,
                    timestamp,
                    tenant.id(),
                    actorUserId,
                    REQUEST_ID,
                    signature,
                    tenant.id() + ":" + REQUEST_ID,
                    canonicalRequest,
                    headers
            );
        }
    }
}
